File: Balanced Binary Tree.cpp
#include<iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };

int Treeheight(TreeNode *root)
    {
        if(root==NULL) return 0;
        else return max(Treeheight(root->left)+1,Treeheight(root->right)+1);
    }

bool isBalanced(TreeNode *root)
    {
        if(root==NULL) return true;
        int len1=Treeheight(root->left),len2=Treeheight(root->right);
        if(max(len1,len2)-min(len1,len2)>1) return false;
        return isBalanced(root->left)&&isBalanced(root->right);
    }

File: Java/Balanced-Binary-Tree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

// time complexity O(n), space complexity O(1)

public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: True if this Binary tree is Balanced, or false.
     */
    public int getDepth(TreeNode root) {
    	if (root == null) return 0;
    	int left = getDepth(root.left), right = getDepth(root.right);
    	if (left == -1 || right == -1 || Math.abs(left - right) > 1) return -1;
    	return max(left, right) + 1;
    }
    public boolean isBalanced(TreeNode root) {
        // write your code here
        return getDepth(root) != -1;
    }
}
